### ISSUE TITLE
T-Clock Startup Speed

### ISSUE THREAD

**Issue by gaveitatry** [White-Tiger/T-Clock#150]
I was just wondering if there is a way to speed up T-Clocks startup.  If I include the T-Clock and the Windows sound icon, I have seven icons on my System Tray that are visible.  And around five others that are hidden that I don't pay attention to because they are hidden.  And T-Clock is usually always the sixth icon out of the seven icons to load.  Is there a way to make it top priority so it gets loaded before all the other icons?

**Comment by White-Tiger** [comment]
What do you mean? As far as I know, T-Clock doesn't have a tray icon... what for anyway?

**Comment by gaveitatry** [comment]
Jesus C.  Let's not play semantics now.  I mean the time it takes to switch from the default computer clock to the T-Clock one.  Since T-Clock is a startup process, I was wondering if there was a way to make it the most top high priority one so it boots up faster.  What are you even here for?  To shut people up and try to get rid of them.  Or to actually help?

**Comment by White-Tiger** [comment]
Well, you said it that way and I've been wondering what you meant, so I had to ask instead of guessing "random" things.
T-Clock has no speed issues that I'm aware of, it's in fact quite fast since it's not doing that much. Yet that doesn't mean that things are working without delays and independent from other factors.
It took ~9 seconds on my Windows 10 VM (so it's slow) before T-Clock showed up after the Explorer Taskbar was visible. While it might be "nice to have" for it to appear instantly, I'm not really seeing an issue here.
I assume (don't know for sure) the current way T-Clock starts with Windows is the slowest of all because it is started by Explorer through the Startup folder. But it doesn't mean that other ways would be way faster or that Windows still doesn't run "everything" else first.
The current way is probably better for your overall Windows startup, because T-Clock wouldn't have to search for the Taskbar and loop around.

How long does it take yours?
Why is it important to see the clock earlier?

> To shut people up and try to get rid of them.

one could say that it's the top priority ;)

**Comment by gaveitatry** [comment]
I don't need to tell you "why is it important".  I hate it when people do that.  When you ask for help and they just dismiss it as "not important" or "I don't understand why you want that".

I like the current method of it starting from the Startup folder.  I have other programs that start from the Startup folder.  I was just wondering if there is a way to get T-Clock to keep starting from the Startup folder but to start as fast or faster than the other programs that start from the Startup folder.

I noticed in Task Manager > Startup, that T-Clock is low priority when starting up.  And some other programs are High or Medium quality.  I am having a difficult time in finding out how to bump T-Clock to be high priority.  I'm not even sure if that would make a difference, but it is worth a try.

So far, I am only getting two uses out of T-Clock.  One is that it makes my clock text white.  The second is that I went ahead and added the weekday above the time.  I'm not using any of the other features like Alarms, Timers, Stopwatches, OnMouseHover, etc.  I'm just trying to get the most out of T-Clock.  I know there is probably a registry tweak that I can do to make the clock text white without having to use T-Clock.  So really the only thing I am getting from T-Clock is the additional weekday information.  If I want to be fully happy with T-Clock, it would need to display white text as soon as I sign in from restart (or after having signed out).  But since you seem unable or incapable of helping me with that, then why bother posting here?

**Comment by blackcrack** [comment]
White Tiger it is not committed for give Support, he can also to close the issue or make it permanently read only, you have maybe give meaningful issue or maybe meaningful suggestion, but bound to give support it is he not, so, he is in anytime in the position for close some issue. This programm it is long enough developed to know where it is a issue or a problem in a windows/NT system and he can say if is important or not, not you have to decide that, but the developer, what is important to him.

I am in favour of this tread/issue being closed and resolved .

best regards
Blackcrack

**Comment by gaveitatry** [comment]
I don't give a crap if you close the issue or not.  I would expect as much from lowlifes.  I have learned long ago to never get my hopes up that human beings would actually act like decent human beings.  I am also an expert on reading people.  I can read them like x-ray vision.  So I already predicted a few issues and a few days ago, that things would end up like this.  I can just tell what kind of person this White-Tiger is and the fact that "someone else" is making it their business to be a scumbag on White-Tiger's behalf just proves what kind of people are behind this github.  Now go away scum.  I am not going to waste any more time with fake people who lack integrity and just want to BS around and play troll games.

**Comment by White-Tiger** [comment]
>  When you ask for help and they just dismiss it [**...**]

asking “why you want that?” is quite important to understand the issue and find a solution. That is, it doesn't have to be always related to one software. Or there's a use case one doesn't see, but which is important nonetheless ;)

> I noticed in Task Manager > Startup, that T-Clock is Low Priority [**...**]

well, that's not the priority.. it's how much it impacts your system's startup.. that is: how long it takes for the app to start and how much resources it uses. Lower is actually better ;)

If most of your other apps use the Startup folder, you could try to rename `T-Clock Redux` to something like: `,T-Clock Redux` _(there's a comma at the beginning)_ or prefix it with `00-`
*(it's in `%AppData%\Microsoft\Windows\Start Menu\Programs\Startup`)*
Otherwise, you could try to use the Registry instead, by adding an entry for T-Clock at:
`HKEY_CURRENT_USER\Software\Microsoft\Windows NT\CurrentVersion\Windows\Run` _(I can create you a .reg file to import if you give me your T-Clock's location)_

> I am in favour of this tread/issue being closed and resolved.

speaking of which... I could close your ReactOS issue xD _(well, at least trying to run T-Clock crashes the entire OS... so wtf?! I'm not sure if I want to compile and fix ReactOS before I can take a look at T-Clock)_

**Comment by blackcrack** [comment]
White-Tiger, hey, it was only a suggestion, and the Reactos-thing is not the theme there, you can close it, no problem .. no, i make it by self.. I just wanted to point that out, but if you shoot now to me, if i stay behind you.. oookey, dann schl135 1ch d13 Sch3155e..

**Comment by gaveitatry** [comment]
Thanks.  It only took six replies to get a decent answer, but at least I got one.  I'll try that stuff later.  And you could have left the BS about "why you want it?" out.  Every single one of your useless posts dealt with that, and it has to be the most stupid question in regards to this particular issue.  If there is a delay of 9 or so seconds and I want to lower the delay, that is all you need to know.  You don't need to know "why".  Although, you almost did a 180 and gave a satisfactory answer, I really don't want anything to do with you guys anymore.  So thank you and good bye.  And Blackcrack, your ugly face wants to close issues for this reason or that reason, and try to make it look like you are a White Knight that only wants to enforce rules... yet you hijack my issue to discuss a completely different and unrelated issue.  So not only are you scum, you are an inconsiderate hypocrite.

**Comment by White-Tiger** [comment]
> yet you hijack my issue

he didn't :P I did^^

And you may call my previous answer a “decent” answer, but every other answer lead to that. Basically you didn't mention the Task Manager before. And of course I could try to explain Windows' startup process, but the "average" user is usually not interested _(and I would to work it out myself first)_... So I didn't talk much about "priorities" there as there barely is one.

I'm still interested into looking into this issue though... but I can't do much unless I get more data / results from you. So behavior might improve by using different startup approaches and trying to influence the startup order through the filename order but I wouldn't know without tests which mostly depend on your setup.

**Comment by gaveitatry** [comment]
> Otherwise, you could try to use the Registry instead, by adding an entry for T-Clock at:
HKEY_CURRENT_USER\Software\Microsoft\Windows NT\CurrentVersion\Windows\Run (I can create you a .reg file to import if you give me your T-Clock's location)

I didn't try this because you didn't provide enough information.  Usually when I find a Registry tweak guide online, they give me full details about what keys to create, etc.  I already know how to get T-Clock's location.  I can just create a shortcut to the Desktop and then copy the path from Properties.  But what about the other stuff?  For example, a recent guide showed me how to create a key somewhere and then add a MUIVerb, SubCommands, Icon, and Postion.  And then in a different key reference the (Default) to something and add a command key.  They were very specific.  So I could follow the instructions immediately with no problem.  You left me guessing in the wind.

> If most of your other apps use the Startup folder, you could try to rename T-Clock Redux to something like: ,T-Clock Redux (there's a comma at the beginning) or prefix it with 00-
(it's in %AppData%\Microsoft\Windows\Start Menu\Programs\Startup)

This actually had a tremendous positive impact.  As mentioned before, I had four other programs plus the sound icon and the NVIDIA icon load before the T-Clock.  That was like six programs loading before T-Clock.  But now, by renaming the Startup shortcut to ,T-Clock, only the sound icon and the NVIDIA icon load before the T-Clock.  It's way faster.  So that helped a lot.

But is there any way to get T-Clock to load faster than the sound icon and the NVIDIA icon without delaying the sound icon and the NVIDIA icon?  Maybe change the , before T-Clock to something else?  Or am I 100% stuck with these startup times?

**Comment by White-Tiger** [comment]
well.. running before NVIDIA's icon is impossible for now. That's because it's using a service and a shell extension to display the icon. So it's not only "global" but does unnecessarily impact the system performance and stability in a negative way.

However, your sound icon is likely another "global" entry in `HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion\Run` which means it is run for all users. So you'd have to create a new "String Value" there, name it in a way that it's above other entries. _(eg. `,T-Clock`, `00-T-Clock`)_ and modify it to set the value data to T-Clock's exe path.
Basically: `HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion\Run` -> Right-click -> New -> String Value -> `,T-Clock` -> <kbd>Return</kbd> -> [Path to Clock.exe]

**Comment by White-Tiger** [comment]
since you've closed the issue, may I assume that it's more or less working for you?
In that case, how much faster does T-Clock start now? So how much time did you save that way?
_(an rough estimate is enough for me, just a number you think is right)_

**Comment by gaveitatry** [comment]
You answered the question about the NVIDIA icon.
What you wrote about the sound icon, but I think you were saying that the sound icon pretty much behaves almost identically to the NVIDIA icon.
Although before your post, I did find an article that says you can change how the Sound icon loads.  Either automatic, manual, disabled, or delayed.  I haven't tried it yet, but you just go to HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\services\Audiosrv.  Make sure the Start dword key is set to 2.  Then create a new dword, name it DelayedAutoStart and set it to 1.  When you want to remove the delya, change DelayedAutoStart to 0.  Source: https://computerstepbystep.com/windows_audio_service.html  I didn't try yet and I don't know if adding a delay will make T-Clock start any faster.
As for what you wrote about HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion\Run, I don't know if I should try it or not.  I mean I already renamed the Startup folder T-Clock shortcut and it was an immediate and noticeable gain.  Will adding a dword entry to the HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion\Run make it even faster?  Or is that an exact equivalent of having a Startup folder shortcut?  I don't want to do something that is redundant with exactly the same effect.
As for second and time and all that stuff.  I never took a stopwatch out and measured it.  I can tell you what it "FEELS" like.  Before I renamed the Startup folder shortcut, it felt like 8-10 seconds.  Now it feels like 2-3 seconds.  So it is a very noticeable improvement.  Like I said, I haven't tried the HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion\Run yet, but I have my doubts that it will speed it up 1-3 seconds faster.
I think before I try these two things, delaying the Sound startup or adding a CurrentVersion\Run entry for T-Clock, I want to see what would happen if I disable T-Clock and try to get a white font clock by just editing the system clock via registry.  I want to see how much faster it will be than T-Clock.  Then, if it is noticeably faster (i.e. 1-3 seconds), then I will go back to using T-Clock and try that Sound delay and CurrentVersion\Run stuff.

**Comment by White-Tiger** [comment]
I'm now confused about which "sound icon" you meant... I assumed you were talking about some 3rd-party icon like Realtek and VIA. These usually use `HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion\Run` as they want their control panel to run for all users.

The "cleaned up" _(and relevant)_ startup order is like this:
1. Services
2. Explorer
   1. Shell extensions _(as they're immediately loaded with Explorer.exe)_
   2. Registry: `HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion\Run`
   3. Registry: `HKEY_CURRENT_USER\Software\Microsoft\Windows\CurrentVersion\Run`
   4. Folder: `%SystemDrive%\Documents and Settings\All Users\Start Menu\Programs\Startup`
   6. Folder: `%AppData%\Microsoft\Windows\Start Menu\Programs\Startup`
3. Probably delayed Services and logon Tasks, though I guess it's possible that Tasks load earlier, maybe even before Explorer)
_(see; https://docs.microsoft.com/en-us/previous-versions/windows/it-pro/windows-xp/bb457123(v=technet.10)#logon-phase and scroll down a bit)_

**Comment by gaveitatry** [comment]
I did not know that the Registry method was faster than the Startup method.  I wish you would have told me that a dozen posts ago.  I would have used that method instead.

I did not know that LOCAL_MACHINE is faster than CURRENT_USER or that All Users is faster than the main User (or whatever you call it).  That is some good information.

I am confused.  Your last paragraph says that creating a Logon Task is the very fastest way.  **I think you mean go to Task Scheduler and create a Task there?**  I have only done that once, like two months ago, but I wrote down the step by step instructions somewhere so I should be able to do that.  The reason I am confused is... well for a lot of reasons.  In your second category 2.) Explorer and Logon Tasks, you clearly listed Shell Extensions (which I don't think is the same thing as a Task Scheduler task), Registry, and then Folders last.  So by that list, since I don't know what  a Shell extension is and probably can't create one, I would think that the Registry method is the very fastest.  But then your last paragraph says the Logon Task is the fastest way.  So you have contradictory information.  Furthermore, I am only guessing that Logon Tasks involve Task Manager.  I am taking a shot in the dark.  It's too bad that you just didn't say it.  I don't want to rely on some CMD you create.  I don't know how safe it is, and I'd rather know how to do something than rely on someone to do it for me.

And this is the second time that you wrote "I assumed".  Why are you always assuming stuff?  If I had a third party Sound program that was launching at Startup, I probably would have told you as much.  But I meant Windows sound.  As in when you have a fresh install of Windows 8 or Windows 10 or probably Windows 7, Vista, and XP as well, you have a sound icon next in the System Tray.  You can do things like access Playback devices, Recording devices, and Volume Mixer.  Why, after a dozen posts, you still want to believe that I have some OTHER sound icon is bewildering to me.

I don't get why All Users loads faster than the main User.  Usually any program I install, I always install it for the main User only and not All Users.  As a result, I have nothing at all in the two following locations:
"C:\ProgramData\Microsoft\Windows\Start Menu\Programs\StartUp"
"C:\ProgramData\Microsoft\Windows\Start Menu\Programs"
It feels strange to me that you are suggesting that I change registries or add Startup shortcuts for an area on my C:\ drive that I never use.

**Comment by White-Tiger** [comment]
>  I wish you would have told me that a dozen posts ago.

I actually did somewhat mention that the registry is "faster" when I said: “If most of your other apps use the Startup folder [...] **Otherwise**, you could try to use the Registry instead, [...]”. Yet you seemed to like the Startup folder, so why not use it if everything uses it anyway?

> I don't get why All Users loads faster than the main User.

"faster" is always in relation to other startup entries.. that is, if neither the `HKEY_LOCAL_MACHINE` registry, nor the `All Users` folder got entries in it, your other apps start right away.
It's just the start order of things when Explorer starts stuff and it tries to wait for them to complete their startup before it goes on. This is to improve the startup speed and responsiveness on HDD drives. _(and maybe for dependencies)_ On SSDs it might be less of a concern... but then again, apps start faster and thus others don't have to wait as long.

> It feels strange to me that you are suggesting that I change registries or add Startup shortcuts for an area on my C:\ drive that I never use.

Which I never did, but you basically asked for "all" the information so I gave it to you.
For the registry part: I bet at least one program placed itself there.
Regarding the `All Users` startup folder: well it's not surprising that none of your programs use it. They usually use the Registry for that.

> The reason I am confused is...

I've put Explorer and Tasks that run at logon into point "2." because I don't know the exact start order and they seem to run at the same time.
The items below that, are started by Explorer, that's why it's a sub-list.

> [...] since I don't know what a Shell extension is and probably can't create one, [...]

"Shell extensions" are DLL files directly loaded by every Explorer process that starts and are things like - **file overlays** _(TortoiseGit, TortoiseSVN, Link Shell Extension, Overdrive/Dropbox icons)_, - **AVs** _(for "Scan this file", "Run in Sandbox", "Scan folder" etc.; they don't have to, but some use shell extensions nonetheless)_, - **Context menu extensions** _(such as "FileMenu Tools")_, - **File archiver** _(eg. 7zip)_, - and other uses

> I don't want to rely on some CMD you create.

well, you would at least learn how the non-GUI way works for creating new Tasks xD
Like I said earlier, I prefer the user-friendly way instead of trying to provide step-by-step instruction which usually fail one way or another.
Anyway, you're not wrong by saying you could use the Task Scheduler to create a logon task.
* Open Task Scheduler:
  * Right-click on `This PC` in an Explorer window or the Start menu _(<Windows 8)_ and choose `Manage`
   Or for Windows 10: <kbd>Shift</kbd>+Right-click Windows icon and choose `Computer Management`
  * Select or Double-click `Task Scheduler`
* Click on `Create Basic Task...` on the right, or Right-click `Task Scheduler` and click on `Create Basic Task...`
* Name: T-Clock
* Trigger: When I log on
* Action: Start a program
* Program/script: <path-to-T-Clock's-Clock.exe> _(or Clock64 on a 64 bit system to start it slightly faster)_
